File: js/complete/no017.js
#!/usr/bin/env node

/* If all the numbers from 1 to 1000 (one thousand) inclusive were written
out in words, how many letters would be used?

NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and
forty-two) contains 23 letters and 115 (one hundred and fifteen) contains
20 letters. The use of "and" when writing out numbers is in compliance
with British usage.

Assume less than or equal to 1000 */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function words(n) {
    var tens = {2: 'twenty', 3: 'thirty', 4: 'forty', 5: 'fifty', 6: 'sixty',
                7: 'seventy', 8:  'eighty', 9: 'ninety'},
        teens = {10: 'ten', 11: 'eleven', 12: 'twelve', 13: 'thirteen',
                 14: 'fourteen', 15: 'fifteen', 16: 'sixteen', 17: 'seventeen',
                 18: 'eighteen', 19: 'nineteen'},
        ones = {1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five',
                6: 'six', 7: 'seven', 8: 'eight', 9: 'nine'};
    if (n == 1000) {
        return 'one thousand';
    }

    var result = [], digs = fns.zeroPad(n, 3).split('');

    if (digs[0] != '0') {
        if (n != 100 * parseInt(digs[0])) {
            result = result.concat([ones[digs[0]], 'hundred', 'and']);
        } else {
            return ones[digs[0]] + ' hundred';
        }
    }

    if (digs[1] == '1') {
        result.push(teens[10 * parseInt(digs[1]) + parseInt(digs[2])]);
        return result.join(' ');
    } else if (digs[1] != '0') {
        result.push(tens[digs[1]]);
    }

    // Here we can safely ignore teens since we return in that loop
    if (digs[2] != '0') {
        result.push(ones[digs[2]]);
    }

    return result.join(' ');
};

function numLettersInWord(n) {
    var result = words(n);

    function spaceHyphenFilter(str) {
        return (str != ' ' && str != '-');
    };    

    // Filter out all with space or hyphens
    result = result.split('').filter(spaceHyphenFilter).join('');
    return result.length;
};

exports.main = function(verbose) {
    if (typeof verbose == 'undefined') {
        verbose = false;
    }
    return operator.sum(operator.range(1, 1001).map(numLettersInWord));
};

if (require.main === module) {
    timer.timer(17, exports.main);
}

File: java/complete/no017.java
class no017 {
  public static String join(String join_val, String[] arr) {
    String result = "";
    int length = arr.length;
    for (int i = 0; i < length - 1; i++) {
      result += arr[i] + join_val;
    }

    if (length > 0) {
      result += arr[length - 1];
    }

    return result;
  }

  public static String words(int n) {
    String[] tens = new String[] {"zero_index_filler", "one_index_filler",
                                  "twenty", "thirty", "forty", "fifty", "sixty",
                                  "seventy",  "eighty", "ninety"};
    String[] teens = new String[] {"ten", "eleven", "twelve", "thirteen",
                                   "fourteen", "fifteen", "sixteen", "seventeen",
                                   "eighteen", "nineteen"};
    String[] ones = new String[] {"zero_index_filler", "one", "two", "three",
                                  "four", "five", "six", "seven",
                                  "eight", "nine"};

    if (n == 1000) {
      return "one thousand";
    }

    String result = "";
    int dig_units = n % 10;
    int dig_tens = ((n - dig_units) / 10) % 10;
    int dig_hundreds = ((n - 10 * dig_tens - dig_units) / 100) % 10;

    if (dig_hundreds != 0) {
      if (n != 100 * dig_hundreds) {
        result += " " + ones[dig_hundreds] + "hundred and";
      } else {
        return ones[dig_hundreds] + " hundred";
      }
    }

    if (dig_tens == 1) {
      result += " " + teens[dig_units]; // index ignores parts ten of n
      return result.trim(); // may have leading space
    } else if (dig_tens != 0) {
      result += " " + tens[dig_tens];
    }

    // Here we can safely ignore teens since we return in that loop
    if (dig_units != 0) {
      result += " " + ones[dig_units];
    }

    return result.trim(); // may have leading space
  }

  public static int num_letters_in_word(int n) {
    String result = words(n);
    result = join("", result.split(" "));
    result = join("", result.split("-"));
    return result.length();
  }

  public static int main(boolean verbose) {
    int result = 0;
    for (int i = 1; i <= 1000; i++) {
      result += num_letters_in_word(i);
    }
    return result;
  }

  public static int main() {
    return main(false);
  }

  public static void main(String[] args) {
    System.out.println(main(true));
  }
}
